### PR TITLE
C++: Revert "Merge pull request #12125 from jketema/unique-function"

### PR DESCRIPTION
**Pull request by MathiasVP**
This reverts commit 9c039c4a080ce4847107ac0012ffebc90df8b42a, reversing changes made to ecd2003c14c9877e0133d398e4fd5635198d8b7d.

### CODE REVIEW

File: cpp/ql/lib/semmle/code/cpp/controlflow/internal/ConstantExprs.qll [hunk 1]
```diff
@@ -110,8 +110,8 @@ private predicate loopConditionAlwaysUponEntry(ControlFlowNode loop, Expr condit
  * should be in this relation.
  */
 pragma[noinline]
-private predicate isFunction(@element el) {
-  el instanceof @function
+private predicate isFunction(Element el) {
+  el instanceof Function
```
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/controlflow/internal/ConstantExprs.qll:114]: I think these changes should just be kept.

File: cpp/ql/lib/semmle/code/cpp/controlflow/internal/ConstantExprs.qll [hunk 2]
```diff
@@ -122,7 +122,7 @@ private predicate isFunction(@element el) {
  */
 pragma[noopt]
 private predicate callHasNoTarget(@funbindexpr fc) {
-  exists(@function f |
+  exists(Function f |
```
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/controlflow/internal/ConstantExprs.qll:125]: Should be kept.

File: cpp/ql/lib/semmle/code/cpp/internal/ResolveGlobalVariable.qll [hunk 3]
```diff
@@ -24,8 +24,8 @@ private predicate isGlobalWithMangledNameAndWithoutDefinition(@mangledname name,
  * a unique global variable `complete` with the same name that does have a definition.
  */
 private predicate hasTwinWithDefinition(@globalvariable incomplete, @globalvariable complete) {
-  not variable_instantiation(incomplete, complete) and
   exists(@mangledname name |
+    not variable_instantiation(incomplete, complete) and
     isGlobalWithMangledNameAndWithoutDefinition(name, incomplete) and
```
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/internal/ResolveGlobalVariable.qll:29]: Should be kept.

Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/internal/ResolveGlobalVariable.qll]: Fixed in fe2c806c2db5f3dceaf2877f34a25fcbce34337b

File: cpp/ql/lib/semmle/code/cpp/controlflow/internal/ConstantExprs.qll (hunk 2 above)
Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/controlflow/internal/ConstantExprs.qll]: Fixed in fe2c806c2db5f3dceaf2877f34a25fcbce34337b

File: cpp/ql/lib/semmle/code/cpp/controlflow/internal/ConstantExprs.qll (hunk 1 above)
Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/controlflow/internal/ConstantExprs.qll]: Fixed in fe2c806c2db5f3dceaf2877f34a25fcbce34337b

**Pull request merged**